### DISCUSSION TITLE
IContentManager vs. ISession

### DISCUSSION THREAD

**Discussion by rjpowers10** [OrchardCMS/OrchardCore#9750] [Q&A] [answered]
My application needs some queries for content items. In old Orchard the `IContentManager` interface provided a way to construct whatever query you needed. In OC `IContentManager` still exists, but from what I can tell the way to write queries now is to use `ISession` directly. This hadn't been a problem for us until I ran into some interesting behavior, and I'm wondering if this is a hole or if I just need to change my code.

Basically, it all comes down to [ContentPartHandlerCoordinator.LoadingAsync](https://github.com/OrchardCMS/OrchardCore/blob/0aa0f4a1e977d8c7624368797bd4971211382dca/src/OrchardCore/OrchardCore.ContentManagement/Handlers/ContentPartHandlerCoordinator.cs#L245). It looks like this method will ensure that all content parts are welded to the content item. My problem is that this loading handler is never called because I am using `ISession` to retrieve my content item instead of `IContentManager`. Had I used [DefaultContentManager.GetAsync](https://github.com/OrchardCMS/OrchardCore/blob/0aa0f4a1e977d8c7624368797bd4971211382dca/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L219) then the loading events would have fired, but I don't know the exact contentItemId so I need to write a query instead. Since the content parts aren't welded to the content item, the `ContentItem.Get` method returns null for a content part that I expect to be found.

Basically, my question is if by retrieving content items directly with `ISession` instead of `IContentManager` I'm leaving the door open for subtle bugs. Should I

- Call `IContentManager.LoadAsync` after I retrieve the content item?
- Call `IContentManager.GetAsync` to retrieve the same content item I just retrieved with `ISession` (to ensure LoadAsync is called)?
- Something else?

**Comment by Piedone** [comment]
You're correct that you should call `IContentManager.LoadAsync()` for each content item you retrieve view `ISession`. And `ISession` queries are the way to fetch content items with any more complex criteria than by IDs. If you do only want to retrieve a single or multiple content items by their IDs then you can use `IContentManager.GetAsync()` and not have to do anything else.

**Reply by rjpowers10** [reply]
This answer makes sense to me but seems rather unfortunate. It's very easy to query for a content item with `ISession` and forget to call `IContentManager.LoadAsync`, which as I noted above can lead to subtle bugs.

**Reply by Piedone** [reply]
I agree, see: https://github.com/OrchardCMS/OrchardCore/issues/5823

**Comment by rjpowers10** [comment, marked as answer]
After reviewing #9161 I think the answer is to make use of `ContentQueryExtensions` and `ContentQueryOfTIndexExtensions`. It's a bit unfortunate but we'll just have to remember to always pass in the content manager when we're getting the query results.
